### PR TITLE
[SPARK-49519][SQL] Merge options of table and relation when constructing FileScanBuilder

### PR DESCRIPTION
**Pull request by liujiayi771**
### What changes were proposed in this pull request?
Merge the options of both `DataSourceV2Relation` and `FileTable` when constructing `FileScanBuilder`.

### Why are the changes needed?
Currently, the subclass of `FileTable` only uses the options from relation when constructing the `FileScanBuilder`, which leads to the omission of the contents in `FileTable.options`. For the `TableCatalog`, the `dsOptions` can be set into the `FileTable.options` returned by the `TableCatalog.loadTable` method. If only the relation options are used here, the `TableCatalog` will not be able to pass `dsOptions` that contains table options to `FileScan`.

Merge the two options is a better option.

### Does this PR introduce _any_ user-facing change?
No.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by liujiayi771** [comment]
Hi @cloud-fan. Could you help to review?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/csv/CSVTable.scala [hunk 1]
```diff
@@ -37,8 +37,13 @@ case class CSVTable(
     userSpecifiedSchema: Option[StructType],
     fallbackFileFormat: Class[_ <: FileFormat])
   extends FileTable(sparkSession, options, paths, userSpecifiedSchema) {
-  override def newScanBuilder(options: CaseInsensitiveStringMap): CSVScanBuilder =
-    CSVScanBuilder(sparkSession, fileIndex, schema, dataSchema, options)
+  override def newScanBuilder(options: CaseInsensitiveStringMap): CSVScanBuilder = {
+    val finalOptions = this.options.asCaseSensitiveMap().asScala.filterNot {
+      case (k, _) => options.containsKey(k)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/csv/CSVTable.scala:42]: do we need this? I think combining two maps will respect the keys in the later map.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/csv/CSVTable.scala [hunk 2]
```diff
@@ -37,8 +37,13 @@ case class CSVTable(
...
+    }.toMap ++ options.asScala.toMap
+    CSVScanBuilder(sparkSession, fileIndex, schema, dataSchema,
+      new CaseInsensitiveStringMap(finalOptions.asJava))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/csv/CSVTable.scala:45]: do other formats have the same issue?

Comment (liujiayi771) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/csv/CSVTable.scala:45]: @cloud-fan orc/parquet/json/textTable do not combine the options of table and relation, while jdbcTable does merge them. I think we can add merging logic for all of them.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
Is it possible to add a test?

**Review by dongjoon-hyun — COMMENTED** [review]
As mentioned above, it needs a test case for the following claim. Please add it to protect your contribution from any potential future regressions, @liujiayi771 .

>  If only the relation options are used here, the TableCatalog will not be able to pass dsOptions that contains table options to FileScan.

**Comment by liujiayi771** [comment]
Okay, I will add a test case, thanks.

**Comment by liujiayi771** [comment]
@cloud-fan @dongjoon-hyun I have added a test case, could you help to review?

**Comment by adrian-wang** [comment]
+1, LGTM

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @liujiayi771 , @cloud-fan , @adrian-wang .
Merged to master for Apache Spark 4.0.0-preview2.

**Comment by dongjoon-hyun** [comment]
To @liujiayi771 , FYI, Apache Spark 4.0.0-preview2 RC1 will be released next Monday.

**Pull request closed without merging**